src/components/modals/MessageSearchModal.js

```diff
@@ -0,0 +1,127 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+
+import { Modal, Popover, OverlayTrigger } from 'react-bootstrap';
+import FaInfoCircle from 'react-icons/lib/fa/info-circle';
+
+import MiniSearch from 'minisearch';
+
+import Message from '../chat/Message';
+
+const searchInfoPopover = (
+  <Popover id="search-info-popover" title="Search Message History">
+    Your messages are encrypted in transit and in storage on our servers. So 
+    how can you search them?
+    <br /><br />
+    Simple! The search happens entirely in the browser. So you can rest easy.
+  </Popover>
+);
+
+
+class MessageSearchModal extends Component {
+  constructor(props) {
+    super(props);
+
+    const miniSearch = this.indexAllMessages(props.messages);
+
+    this.state = {
+      username: this.props.username,
+      isVisible: this.props.isVisible,
+      miniSearch: miniSearch,
+      isLoadingResults: false,
+      searchResults: []
+    };
+  }
+
+  indexAllMessages(messages) {
+    const miniSearch = new MiniSearch({
+      fields: ['msg', 'from'], // fields to index for full-text search
+      storeFields: ['msg', 'from'] // fields to return with search results
+    });
+
+    messages = this.props.messages.map((message, i) => {
+      return {"id": i, ...message};
+    });
+    miniSearch.addAll(messages);
+
+    return miniSearch;
+  };
+  
+  getSearchResults = (e) => {
+    this.setState({
+      isLoadingResults: true
+    });
+    let searchResults = this.state.miniSearch.search(e.target.value);
+    this.setState({
+      searchResults: searchResults,
+      isLoadingResults: false
+    });
+  }
+
+  onClose = () => {
+    // belt and suspenders; explicitly wipe local index from state
+    this.setState({ miniSearch: null });
+    this.props.onCloseModal();
+  }
+
+  render() {
+    const {
+      isVisible,
+      searchResults,
+      isLoadingResults,
+      username } = this.state;
+
+    const sortedResults = searchResults.sort((a, b) => a.score > b.score);
```

Comment:
@jimmcgaw Hey hey.

This line mutates `searchResults` so it may not make sense to create `sortedResults` as a separate variable.

Reply:
Oh yeahhh, I forgot that this method sorts an array in-place. I’ll remove the extra car.